File: notes/2026-07-17-c260-a5-template-nimber-crosscheck.py
#!/usr/bin/env python3
"""C260 independent cross-check of A5 (and S4/V4) regular-template Node-Kayles nimbers.

Independent recomputation of the free-orbit (regular template) Node-Kayles Grundy values
that otherwise rest on the single Rust solver rust/scripts/nodekayles_cayley.rs.

INDEPENDENCE (see companion .md for the full statement):
  * Necessarily SHARED: the graph definition. For a generating involution triple T of a
    group G acting regularly, the template is the cubic Cayley graph Cay(G,T): one vertex
    per group element, an edge {g, g*t} for each t in T. This script rebuilds that graph in
    its own code, but the mathematical object is the same one the Rust solver evaluates.
  * INDEPENDENT: (a) the solver code here is freshly written; (b) the memo canonicalization
    is graph-isomorphism canonical labelling of each connected induced subposition, computed
    by the BLISS engine inside python-igraph -- NOT the hand-built left-mult x color-perm
    group-orbit minimum used by the Rust solver. Node-Kayles Grundy value is a graph
    isomorphism invariant, so keying the memo on a per-subgraph canonical certificate is
    sound and is a strictly different (and generally stronger) dedup than the Rust path.

The Rust solver and this script are mutually independent replays of the same finite
computation; agreement of all values is the cross-check.

Determinism: enumeration is fixed (no randomness, no seeds). igraph.canonical_permutation()
is deterministic. JSON output is sorted and timestamp-free.

Usage (run from repo rust/ dir, or anywhere):
    uv run --with igraph python3 <thisfile> <target> [sig]
      <target> in {V4, S4, A5}; optional <sig> like 2,3,5 restricts to one class.
    uv run --with igraph python3 <thisfile> emit-json > out.json   # all classes, canonical

Dependencies: python-igraph (BLISS canonical labelling). Pinned version reported in the JSON.
"""
import sys
import json
import time

# ---------------------------------------------------------------------------
# Permutation group machinery. Composition convention (p*q)[i] = p[q[i]] matches
# the Rust solver so the constructed Cayley graphs are identical, not merely iso.
# ---------------------------------------------------------------------------

def mul(p, q):
    return tuple(p[q[i]] for i in range(len(q)))

def is_ident(p):
    return all(i == x for i, x in enumerate(p))

def perm_order(p):
    x = p
    o = 1
    while not is_ident(x):
        x = mul(p, x)
        o += 1
    return o

def sign(p):
    inv = 0
    n = len(p)
    for i in range(n):
        for j in range(i + 1, n):
            if p[i] > p[j]:
                inv += 1
    return 1 if inv % 2 == 0 else -1

def all_perms(n):
    """Same swap-recursion order as the Rust all_perms()."""
    v = list(range(n))
    out = []

    def rec(k):
        if k == n:
            out.append(tuple(v))
            return
        for i in range(k, n):
            v[k], v[i] = v[i], v[k]
            rec(k + 1)
            v[k], v[i] = v[i], v[k]

    rec(0)
    return out

def closure(gens, n):
    ident = tuple(range(n))
    seen = {ident}
    frontier = [ident]
    while frontier:
        g = frontier.pop()
        for s in gens:
            h = mul(g, s)
            if h not in seen:
                seen.add(h)
                frontier.append(h)
    return sorted(seen)

# ---------------------------------------------------------------------------
# Build one representative Cayley graph per pairwise-product-order signature.
# Mirrors solve_group() in the Rust solver: first triple (in the fixed element
# ordering) per signature whose closure is the whole group.
# ---------------------------------------------------------------------------

def build_classes(n, even_only):
    target = (factorial(n) // 2) if even_only else factorial(n)
    els = [p for p in all_perms(n) if (not even_only) or sign(p) == 1]
    invols = [p for p in els if (not is_ident(p)) and is_ident(mul(p, p))]
    reps = {}  # sig(tuple) -> triple
    m = len(invols)
    for i in range(m):
        for j in range(i + 1, m):
            for k in range(j + 1, m):
                tri = (invols[i], invols[j], invols[k])
                if len(closure(tri, n)) != target:
                    continue
                sig = tuple(sorted((
                    perm_order(mul(tri[0], tri[1])),
                    perm_order(mul(tri[0], tri[2])),
                    perm_order(mul(tri[1], tri[2])),
                )))
                if sig not in reps:
                    reps[sig] = tri
    return target, invols, reps

def factorial(n):
    r = 1
    for i in range(2, n + 1):
        r *= i
    return r

def cayley_graph(tri, n):
    group = closure(tri, n)
    idx = {g: i for i, g in enumerate(group)}
    v = len(group)
    adj = [0] * v
    for g in group:
        i = idx[g]
        for t in tri:
            h = mul(g, t)
            j = idx[h]
            if i != j:
                adj[i] |= 1 << j
                adj[j] |= 1 << i
    closed = [adj[x] | (1 << x) for x in range(v)]
    return v, adj, closed

def k4_graph():
    """V4 template = K4 = Cay(V4, its 3 involutions). Closed-form Grundy = 1."""
    v = 4
    adj = [0] * v
    for i in range(v):
        for j in range(v):
            if i != j:
                adj[i] |= 1 << j
    closed = [adj[x] | (1 << x) for x in range(v)]
    return v, adj, closed

# ---------------------------------------------------------------------------
# Independent Node-Kayles Grundy solver.
#   * connected-component decomposition, Grundy = XOR of component Grundies
#   * memo keyed on BLISS graph-iso canonical certificate of the induced subgraph
# ---------------------------------------------------------------------------

def bits(mask):
    out = []
    while mask:
        low = mask & (-mask)
        out.append(low.bit_length() - 1)
        mask &= mask - 1
    return out

def make_solver(adj, closed):
    import igraph

    raw_memo = {}
    canon_memo = {}
    stats = {"canon_calls": 0}

    def component(rem, start):
        comp = start
        while True:
            nb = 0
            b = comp
            while b:
                v = (b & -b).bit_length() - 1
                nb |= adj[v]
                b &= b - 1
            new = comp | (nb & rem)
            if new == comp:
                return comp
            comp = new

    def canon(comp):
        stats["canon_calls"] += 1
        verts = bits(comp)
        loc = {v: i for i, v in enumerate(verts)}
        edges = []
        for v in verts:
            nb = adj[v] & comp
            b = nb
            while b:
                w = (b & -b).bit_length() - 1
                if v < w:
                    edges.append((loc[v], loc[w]))
                b &= b - 1
        g = igraph.Graph(n=len(verts), edges=edges)
        perm = g.canonical_permutation()
        ce = sorted(
            (min(perm[a], perm[b]), max(perm[a], perm[b])) for (a, b) in edges
        )
        return (len(verts), tuple(ce))

    def grundy(mask):
        if mask == 0:
            return 0
        g = 0
        rem = mask
        while rem:
            start = rem & (-rem)
            comp = component(rem, start)
            g ^= grundy_conn(comp)
            rem &= ~comp
        return g

    def grundy_conn(comp):
        v = raw_memo.get(comp)
        if v is not None:
            return v
        key = canon(comp)
        v = canon_memo.get(key)
        if v is not None:
            raw_memo[comp] = v
            return v
        opts = set()
        b = comp
        while b:
            vv = (b & -b).bit_length() - 1
            child = comp & ~closed[vv]
            opts.add(grundy(child))
            b &= b - 1
        m = 0
        while m in opts:
            m += 1
        canon_memo[key] = m
        raw_memo[comp] = m
        return m

    def solve(nv):
        full = (1 << nv) - 1
        return grundy(full)

    return solve, canon_memo, raw_memo, stats

def graph_girth(adj, nv):
    from collections import deque
    best = None
    for s in range(nv):
        dist = [-1] * nv
        par = [-1] * nv
        dist[s] = 0
        dq = deque([s])
        while dq:
            u = dq.popleft()
            b = adj[u]
            while b:
                w = (b & -b).bit_length() - 1
                b &= b - 1
                if dist[w] == -1:
                    dist[w] = dist[u] + 1
                    par[w] = u
                    dq.append(w)
                elif w != par[u]:
                    c = dist[u] + dist[w] + 1
                    if best is None or c < best:
                        best = c
    return best

def run_target(target, sig_filter=None, verbose=True):
    results = []
    if target == "V4":
        nv, adj, closed = k4_graph()
        solve, cm, rm, st = make_solver(adj, closed)
        t0 = time.time()
        g = solve(nv)
        dt = time.time() - t0
        girth = 3
        rec = {"group": "V4", "sig": [2, 2, 2], "vertices": nv, "edges": 6,
               "girth": girth, "grundy": g, "canon_states": len(cm),
               "raw_states": len(rm), "seconds": round(dt, 3)}
        results.append(rec)
        if verbose:
            print(f"V4 sig=(2,2,2) K4: G={g} canon={len(cm)} raw={len(rm)} {dt:.2f}s")
        return results
    n = 4 if target == "S4" else 5
    even = (target == "A5")
    _, invols, reps = build_classes(n, even)
    for sig in sorted(reps):
        if sig_filter is not None and list(sig) != list(sig_filter):
            continue
        tri = reps[sig]
        nv, adj, closed = cayley_graph(tri, n)
        edges = sum(bin(adj[i]).count("1") for i in range(nv)) // 2
        girth = graph_girth(adj, nv)
        solve, cm, rm, st = make_solver(adj, closed)
        t0 = time.time()
        g = solve(nv)
        dt = time.time() - t0
        rec = {"group": target, "sig": list(sig), "vertices": nv, "edges": edges,
               "girth": girth, "grundy": g, "canon_states": len(cm),
               "raw_states": len(rm), "seconds": round(dt, 3)}
        results.append(rec)
        if verbose:
            print(f"{target} sig={sig}: G={g} vtx={nv} edges={edges} girth={girth} "
                  f"canon={len(cm)} raw={len(rm)} {dt:.1f}s")
    return results

def emit_json():
    import igraph
    out = {
        "task": "C260",
        "description": "Independent cross-check of A5/S4/V4 regular-template Node-Kayles nimbers",
        "method": "graph-iso canonical (BLISS/igraph) memoized Sprague-Grundy, "
                  "connected-component xor; independent of rust/scripts/nodekayles_cayley.rs",
        "igraph_version": igraph.__version__,
        "convention": "template = Cay(G,T) cubic; move deletes closed nbhd N[v]; "
                      "sig = sorted pairwise product orders of the generating triple",
        "results": [],
    }
    for tgt in ("V4", "S4", "A5"):
        out["results"].extend(run_target(tgt, verbose=False))
    # canonical sort
    out["results"].sort(key=lambda r: (r["group"], r["sig"]))
    for r in out["results"]:
        r.pop("seconds", None)  # timings are not canonical evidence
    print(json.dumps(out, indent=2, sort_keys=True))

def main():
    args = sys.argv[1:]
    if not args:
        print("usage: <target V4|S4|A5> [sig like 2,3,5] | emit-json")
        return
    if args[0] == "emit-json":
        emit_json()
        return
    tgt = args[0]
    sig = None
    if len(args) > 1:
        sig = [int(x) for x in args[1].split(",")]
    run_target(tgt, sig_filter=sig, verbose=True)

if __name__ == "__main__":
    main()

File: notes/2026-07-17-c260-a5-template-nimber-crosscheck.rs
// C260 independent cross-check (fast backstop) of A5/S4 regular-template Node-Kayles nimbers.
//
// This is an INDEPENDENT reimplementation, written fresh for C260, NOT derived from
// rust/scripts/nodekayles_cayley.rs.
//
//   * Necessarily SHARED: the graph definition. The template is the cubic Cayley graph
//     Cay(G,T) (one vertex per group element, an edge {g, g*t} per involution t in T).
//     This program rebuilds it with the same composition convention and the same
//     first-triple-per-pairwise-order-signature selection, so it evaluates the SAME graphs.
//
//   * INDEPENDENT: fresh solver code, and a deliberately DIFFERENT canonicalization -- the
//     memo key is the orbit minimum of a connected component's bitmask under the
//     left-regular representation of G ONLY (|G| permutations). The reference Rust solver
//     additionally minimises over the color-permuting conjugation stabiliser of T, so for
//     every class whose signature has a repeated order (all A5 classes except (2,3,5)) this
//     program uses a strictly SMALLER canonicalization group and therefore performs a
//     genuinely different computation, reaching more memo states and canonicalising them by
//     a different key. (For (2,3,5) that stabiliser is already trivial, so the group
//     coincides and this class is covered independently by the companion Python BLISS
//     per-subgraph-isomorphism solver.)
//
// Left multiplication L_g : h |-> g*h is an automorphism of the right Cayley graph, so
// orbit-min under {L_g} is a sound value-preserving canonicalization. Grundy of a
// disconnected position = XOR of component Grundies (Sprague-Grundy sum).
//
// Build & run (from repo rust/):
//   rustc -O ../notes/2026-07-17-c260-a5-template-nimber-crosscheck.rs -o /tmp/c260nk
//   /tmp/c260nk S4            # 4 classes -> 0
//   /tmp/c260nk A5            # 5 classes
//   /tmp/c260nk A5 5,5,5      # one class by signature
//   /tmp/c260nk json         # canonical JSON of all S4+A5 class values to stdout

use std::collections::HashMap;

fn mul(p: &[u32], q: &[u32]) -> Vec<u32> {
    (0..q.len()).map(|i| p[q[i] as usize]).collect()
}
fn is_ident(p: &[u32]) -> bool {
    p.iter().enumerate().all(|(i, &x)| i as u32 == x)
}
fn perm_order(p: &[u32]) -> u32 {
    let mut x = p.to_vec();
    let mut o = 1u32;
    while !is_ident(&x) {
        x = mul(p, &x);
        o += 1;
    }
    o
}
fn sign(p: &[u32]) -> i32 {
    let mut inv = 0;
    for i in 0..p.len() {
        for j in (i + 1)..p.len() {
            if p[i] > p[j] {
                inv += 1;
            }
        }
    }
    if inv % 2 == 0 {
        1
    } else {
        -1
    }
}
fn all_perms(n: usize) -> Vec<Vec<u32>> {
    // same swap-recursion order as the reference, so element indexing matches
    let mut v: Vec<u32> = (0..n as u32).collect();
    let mut out = Vec::new();
    fn rec(v: &mut Vec<u32>, k: usize, out: &mut Vec<Vec<u32>>) {
        if k == v.len() {
            out.push(v.clone());
            return;
        }
        for i in k..v.len() {
            v.swap(k, i);
            rec(v, k + 1, out);
            v.swap(k, i);
        }
    }
    rec(&mut v, 0, &mut out);
    out
}
fn closure(gens: &[Vec<u32>], n: usize) -> Vec<Vec<u32>> {
    let ident: Vec<u32> = (0..n as u32).collect();
    let mut seen: std::collections::HashSet<Vec<u32>> = std::collections::HashSet::new();
    seen.insert(ident.clone());
    let mut frontier = vec![ident];
    while let Some(g) = frontier.pop() {
        for s in gens {
            let h = mul(&g, s);
            if seen.insert(h.clone()) {
                frontier.push(h);
            }
        }
    }
    let mut v: Vec<Vec<u32>> = seen.into_iter().collect();
    v.sort();
    v
}

fn component(rem: u64, start: u64, adj: &[u64]) -> u64 {
    let mut comp = start;
    loop {
        let mut nb = 0u64;
        let mut b = comp;
        while b != 0 {
            let v = b.trailing_zeros() as usize;
            nb |= adj[v];
            b &= b - 1;
        }
        let new = comp | (nb & rem);
        if new == comp {
            return comp;
        }
        comp = new;
    }
}
#[inline]
fn apply_perm(mask: u64, perm: &[u32]) -> u64 {
    let mut r = 0u64;
    let mut b = mask;
    while b != 0 {
        let v = b.trailing_zeros() as usize;
        r |= 1u64 << perm[v];
        b &= b - 1;
    }
    r
}
#[inline]
fn canonical(mask: u64, perms: &[Vec<u32>]) -> u64 {
    let mut best = mask;
    for p in perms {
        let m = apply_perm(mask, p);
        if m < best {
            best = m;
        }
    }
    best
}
fn grundy(mask: u64, adj: &[u64], closed: &[u64], perms: &[Vec<u32>], memo: &mut HashMap<u64, u8>) -> u8 {
    if mask == 0 {
        return 0;
    }
    let mut rem = mask;
    let mut g = 0u8;
    while rem != 0 {
        let start = rem & rem.wrapping_neg();
        let comp = component(rem, start, adj);
        g ^= grundy_conn(comp, adj, closed, perms, memo);
        rem &= !comp;
    }
    g
}
fn grundy_conn(mask: u64, adj: &[u64], closed: &[u64], perms: &[Vec<u32>], memo: &mut HashMap<u64, u8>) -> u8 {
    let key = canonical(mask, perms);
    if let Some(&v) = memo.get(&key) {
        return v;
    }
    let mut opts = 0u64;
    let mut b = mask;
    while b != 0 {
        let v = b.trailing_zeros() as usize;
        let child = mask & !closed[v];
        let cg = grundy(child, adj, closed, perms, memo);
        opts |= 1u64 << cg;
        b &= b - 1;
    }
    let m = (!opts).trailing_zeros() as u8;
    memo.insert(key, m);
    m
}

struct ClassResult {
    sig: [u32; 3],
    vertices: usize,
    edges: usize,
    grundy: u8,
    memo_states: usize,
}

fn build_reps(n: usize, even_only: bool) -> Vec<([u32; 3], [Vec<u32>; 3])> {
    let target = if even_only { fact(n) / 2 } else { fact(n) };
    let els: Vec<Vec<u32>> = all_perms(n)
        .into_iter()
        .filter(|p| !even_only || sign(p) == 1)
        .collect();
    let invols: Vec<Vec<u32>> = els
        .iter()
        .filter(|p| !is_ident(p) && is_ident(&mul(p, p)))
        .cloned()
        .collect();
    let mut reps: std::collections::BTreeMap<[u32; 3], [Vec<u32>; 3]> = std::collections::BTreeMap::new();
    for i in 0..invols.len() {
        for j in (i + 1)..invols.len() {
            for k in (j + 1)..invols.len() {
                let tri = [invols[i].clone(), invols[j].clone(), invols[k].clone()];
                if closure(&tri, n).len() != target {
                    continue;
                }
                let mut sig = [
                    perm_order(&mul(&tri[0], &tri[1])),
                    perm_order(&mul(&tri[0], &tri[2])),
                    perm_order(&mul(&tri[1], &tri[2])),
                ];
                sig.sort();
                reps.entry(sig).or_insert(tri);
            }
        }
    }
    reps.into_iter().collect()
}

fn solve_class(n: usize, tri: &[Vec<u32>; 3]) -> ClassResult {
    let group = closure(tri, n);
    let idx: HashMap<Vec<u32>, usize> = group.iter().cloned().enumerate().map(|(i, g)| (g, i)).collect();
    let v = group.len();
    let mut adj = vec![0u64; v];
    for g in &group {
        let i = idx[g];
        for t in tri.iter() {
            let h = mul(g, t);
            let j = idx[&h];
            if i != j {
                adj[i] |= 1u64 << j;
                adj[j] |= 1u64 << i;
            }
        }
    }
    let edges: usize = adj.iter().map(|a| a.count_ones() as usize).sum::<usize>() / 2;
    let closed: Vec<u64> = (0..v).map(|x| adj[x] | (1u64 << x)).collect();
    // left-regular representation ONLY (|G| permutations): perm_g[idx[h]] = idx[g*h]
    let perms: Vec<Vec<u32>> = group
        .iter()
        .map(|g| group.iter().map(|h| idx[&mul(g, h)] as u32).collect())
        .collect();
    let sig = {
        let mut s = [
            perm_order(&mul(&tri[0], &tri[1])),
            perm_order(&mul(&tri[0], &tri[2])),
            perm_order(&mul(&tri[1], &tri[2])),
        ];
        s.sort();
        s
    };
    let full = if v == 64 { u64::MAX } else { (1u64 << v) - 1 };
    let mut memo: HashMap<u64, u8> = HashMap::new();
    let val = grundy(full, &adj, &closed, &perms, &mut memo);
    ClassResult {
        sig,
        vertices: v,
        edges,
        grundy: val,
        memo_states: memo.len(),
    }
}

fn fact(n: usize) -> usize {
    (1..=n).product()
}

fn run(name: &str, n: usize, even: bool, filter: Option<[u32; 3]>, json: bool, out: &mut Vec<(String, ClassResult)>) {
    let reps = build_reps(n, even);
    for (sig, tri) in &reps {
        if let Some(f) = filter {
            if *sig != f {
                continue;
            }
        }
        let r = solve_class(n, tri);
        if !json {
            eprintln!(
                "{} sig={:?}: G={} vtx={} edges={} memo_states={}",
                name, r.sig, r.grundy, r.vertices, r.edges, r.memo_states
            );
        }
        out.push((name.to_string(), r));
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mode = args.get(1).map(|s| s.as_str()).unwrap_or("A5");
    let filter: Option<[u32; 3]> = args.get(2).map(|s| {
        let v: Vec<u32> = s.split(',').map(|x| x.trim().parse().unwrap()).collect();
        [v[0], v[1], v[2]]
    });
    let mut out: Vec<(String, ClassResult)> = Vec::new();
    match mode {
        "S4" => run("S4", 4, false, filter, false, &mut out),
        "A5" => run("A5", 5, true, filter, false, &mut out),
        "json" => {
            run("S4", 4, false, None, true, &mut out);
            run("A5", 5, true, None, true, &mut out);
            // canonical JSON: sorted by (group, sig)
            out.sort_by(|a, b| (a.0.clone(), a.1.sig).cmp(&(b.0.clone(), b.1.sig)));
            println!("{{");
            println!("  \"solver\": \"c260-rust-leftmult-only\",");
            println!("  \"results\": [");
            for (i, (g, r)) in out.iter().enumerate() {
                let comma = if i + 1 < out.len() { "," } else { "" };
                println!(
                    "    {{\"group\": \"{}\", \"sig\": [{}, {}, {}], \"vertices\": {}, \"edges\": {}, \"grundy\": {}}}{}",
                    g, r.sig[0], r.sig[1], r.sig[2], r.vertices, r.edges, r.grundy, comma
                );
            }
            println!("  ]");
            println!("}}");
            return;
        }
        _ => {
            eprintln!("usage: <S4|A5|json> [sig like 5,5,5]");
            return;
        }
    }
}
